apache/superset
# fix: Translations related to the date range filter

### SUMMARY
This PR *partially* addresses discussion #25994 & issue #26061. It fixes the issue with the date range filter not showing translated values.

The issue was that the front-end client was sending requests that asked for strings like "Last year" and displaying the `timeRange` property in the response. As it stands now, the backend translates the human-readable string into actual dates, along with the requested string in the `timeRange` property (see [api.py#L98](https://github.com/apache/superset/blob/master/superset/views/api.py#L98). The front-end needed to translate the property coming back, but also needed to be sure that it expected the same value it sent (see changes in DateRangeFilter.tsx).

One thing I will say is that it was not clear at first how to go about fixing translation issues. I initially did the same as #26010, thinking that updating the .po and .json files was the correct path.
It wasn't until later that I realized that a 'proper' fix would follow the instructions [in the docs](https://superset.apache.org/docs/contributing/translations/). Most of this PR is actually the result of trying to resolve the issues that existed in the translation pipeline.

I'm going to outline it here. I know I was confused for a while, and it seems like others were also confused as per the discussion in #26010.
The translation is a multi-step process:
1. One should do a full extraction [/scripts/babel_update.sh](https://github.com/apache/superset/blob/master/scripts/babel_update.sh)
  - This will generate the .pot file based on the strings in the code
  - Then it will merge the .pot file with the existing .po files to the best of its ability
    - This will sometimes cause the tool to make 'guesses' at what the value should be if it doesn't already exist. Those guesses are annotated as 'fuzzy'.
2. The next step is to review/edit/fix the values in the .po file
3. Next generate the .json file(s) with the [/scripts/po2json.sh](https://github.com/apache/superset/blob/master/scripts/po2json.sh) file.
  - The .json file is used by the frontend code for translations
  - 'fuzzy' values in the .po are not translated to the .json files.
4. Finally generate the .mo file(s) with the command `pybabel compile -d superset/translations`
  - The .mo file is used by the backend code for translations
  - The .mo file is *not* checked into the git codebase. I'm unclear as to why the .json file is if the .mo file is not.
  - It is unclear looking at the UI which strings are generated by the frontend and which are generated by the backend

What is not addressed by this PR from the original issue is the time-grain values. I've looked into fixing this, but since the front-end is using standard 'select' input fields to display this, there's no good location in the front-end to force a translation. I do have a branch that does the translation in the backend (which is why I had to dig so deep into the translation stuff for the backend .mo files), but that's breaking some integration tests so I need to find another way to do this. I'm open to suggestions if anyone has any.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Before:
![image](https://github.com/apache/superset/assets/90489859/d78be8f7-d695-4efc-aa40-f4678a62160a)
After:
![image](https://github.com/apache/superset/assets/90489859/a263a8cb-c5cd-49ad-98f9-0dad8fbd37fa)


### TESTING INSTRUCTIONS
1. Launch a new instance of superset loaded with examples and the French language enabled
2. Open the "Birth Rates" dashboard
3. Create a new "Time Range" filter
4. Switch to French
5. [ ] Verify that the new filter display the properly translated values after selecting one of the 'last X' options instead of the English versions

### ADDITIONAL INFORMATION
  - Associated to but does not fix #17779
  - Associated to but does not fix #25728
  - Associated to but does not fix #26061
  - Related to PR #26010
  - Related to Discussion #25994
